sw/host/tests/chip/rv_dm/src/jtag_tap_sel.rs

```diff
@@ -0,0 +1,108 @@
+use clap::Parser;
+use once_cell::sync::Lazy;
+use regex::Regex;
+
+use opentitanlib::app::TransportWrapper;
+use opentitanlib::dif::lc_ctrl::{DifLcCtrlState, LcCtrlReg};
+use opentitanlib::execute_test;
+use opentitanlib::io::jtag::JtagTap;
+use opentitanlib::test_utils::init::InitializeTest;
+
+use top_earlgrey::top_earlgrey;
+
+#[derive(Debug, Parser)]
+struct Opts {
+    #[command(flatten)]
+    init: InitializeTest,
+}
+
+// Needs to match util/openocd/target
+const RISCV_IDCODE: u32 = 0x10001cdf;
+const LC_IDCODE: u32 = 0x10002cdf;
+
+fn test_jtag_tap_sel(opts: &Opts, transport: &TransportWrapper) -> Result<()> {
+    // Avoid watchdog timeout by entering bootstrap mode.
+    transport.pin_strapping("ROM_BOOTSTRAP")?.apply()?;
+
+    // Test the LC TAP
+    transport.pin_strapping("PINMUX_TAP_LC")?.apply()?;
+    transport.reset_target(opts.init.bootstrap.options.reset_delay, true)?;
+
+    let mut jtag = opts
+        .init
+        .jtag_params
+        .create(transport)?
+        .connect(JtagTap::LcTap)?;
+    let lc_state =
+        DifLcCtrlState::from_redundant_encoding(jtag.read_lc_ctrl_reg(&LcCtrlReg::LcState)?)?;
+
+    jtag.disconnect()?;
+    transport.pin_strapping("PINMUX_TAP_LC")?.remove()?;
+
+    // Test RISC-V TAP
+    transport.pin_strapping("PINMUX_TAP_RISCV")?.apply()?;
+    transport.reset_target(opts.init.bootstrap.options.reset_delay, true)?;
+
+    let mut jtag = opts
+        .init
+        .jtag_params
+        .create(transport)?
+        .connect(JtagTap::RiscvTap)?;
+    let mut lc_state_rv = 0;
+    jtag.read_memory32(
+        top_earlgrey::LC_CTRL_BASE_ADDR as u32 + LcCtrlReg::LcState as u32,
+        core::slice::from_mut(&mut lc_state_rv),
+    )?;
+    let lc_state_rv = DifLcCtrlState::from_redundant_encoding(lc_state_rv)?;
+    assert_eq!(lc_state, lc_state_rv);
+
+    jtag.disconnect()?;
+    transport.pin_strapping("PINMUX_TAP_RISCV")?.remove()?;
+
+    // Test DFT TAP
+    transport.pin_strapping("PINMUX_TAP_DFT")?.apply()?;
+    transport.reset_target(opts.init.bootstrap.options.reset_delay, true)?;
+
+    // For DFT test we need to do raw OpenOCD command.
+    let mut openocd = opts.init.jtag_params.create(transport)?.into_raw()?;
+    // Perform JTAG initialisation and capture the result.
+    let init_result = openocd.execute("capture \"jtag init\"")?;
+
+    static ID_REGEX: Lazy<Regex> =
+        Lazy::new(|| Regex::new(r"tap/device found: 0x([0-9A-Fa-f]+) \(").unwrap());
+    let idcode = if init_result.contains("JTAG scan chain interrogation failed") {
+        None
+    } else {
+        Some(u32::from_str_radix(
+            ID_REGEX
+                .captures(&init_result)
+                .context("Failed to parse IDCODE from OpenOCD output")?
+                .get(1)
+                .unwrap()
+                .as_str(),
+            16,
+        )?)
+    };
+
+    if let Some(idcode) = idcode {
+        assert_ne!(idcode, RISCV_IDCODE);
+        assert_ne!(idcode, LC_IDCODE);
+    }
```

Comment:
I feel like this might be a little too loose for real silicon (and unexpectedly not trigger failure on getting no response from any TAP). It seems fine for the FPGA, though! There's no TAP there, so anything that isn't the known IDCODE values seems fine. Technically, it should really be a case of all ones if we're properly pulling up on TDO, but I'm not sure whether the Olimex adapter is doing so.

Would it make sense to you to add an optional argument that, when set (to a DFT IDCODE), would check that the received `idcode` actually matches? Then for the later build with real silicon, we could fill it in for that target.

Reply:
> case of all ones

OpenOCD handles all one case specially and gives you "JTAG scan chain interrogation failed"

Reply:
I've updated the test to accept an optional `--dft-idcode=0x<IDCODE>`. If not specified, then it checks for non-existence of DFT TAP.